### PR TITLE
Set unsupported key for libvirt

### PR DESCRIPTION
**Pull request by praveenkumar**
This PR is an initial step to make single node cluster settings easy for libvirt provider. Currently we need to patch the etcd resource[0] after bootstrap is up and that is a manual step which blocks CRC bundle creation on the CI. 

[0] https://github.com/openshift/cluster-etcd-operator/pull/279

```
$ podman run --volume "$PWD:/assets:z" quay.io/praveenkumar/etcd-operator:latest /usr/bin/cluster-etcd-operator render --etcd-ca=/assets/tls/etcd-ca-bundle.crt  --etcd-metric-ca=/assets/tls/etcd-metric-ca-bundle.crt  --manifest-etcd-image=foobar  --etcd-discovery-domain=testing  --manifest-cluster-etcd-operator-image=barfoo  --manifest-setup-etcd-env-image=hohoho  --manifest-kube-client-agent-image=afjakfja  --asset-input-dir=/assets/tls  --asset-output-dir=/assets/etcd-bootstrap  --config-output-file=/assets/etcd-bootstrap/config  --cluster-config-file=/assets/manifests/cluster-network-02-config.yml
Writing asset: /assets/etcd-bootstrap/bootstrap-manifests/etcd-member-pod.yaml
Writing asset: /assets/etcd-bootstrap/manifests/00_etcd-host-service-endpoints.yaml
Writing asset: /assets/etcd-bootstrap/manifests/00_etcd-host-service.yaml
Writing asset: /assets/etcd-bootstrap/manifests/00_openshift-etcd-ns.yaml
Writing asset: /assets/etcd-bootstrap/manifests/configmap-initial-etcd-serving-ca.yaml
Writing asset: /assets/etcd-bootstrap/manifests/0000_12_etcd-operator_01_operator.cr.yaml

$ cat etcd-bootstrap/manifests/0000_12_etcd-operator_01_operator.cr.yaml  
{
  "kind": "Etcd",
  "apiVersion": "operator.openshift.io/v1",
  "metadata": {
    "name": "cluster",
    "creationTimestamp": null,
    "annotations": {
      "release.openshift.io/create-only": "true"
    }
  },
  "spec": {
    "managementState": "Managed",
    "logLevel": "",
    "operatorLogLevel": "",
    "unsupportedConfigOverrides": {
      "useUnsupportedUnsafeNonHANonProductionUnstableEtcd": true
    },
    "observedConfig": null
  },
  "status": {
    "readyReplicas": 0
  }
}
```

### CODE REVIEW

File: pkg/cmd/render/render.go [hunk 1]
```diff
@@ -377,6 +382,42 @@ func (t *TemplateData) setPlatform(infraConfigFilePath string) error {
...
+	etcdOperatorCRStructure := &operatorv1.MyOperatorResource{}
+	etcdOperatorCRData, err := ioutil.ReadFile(etcdOperatorCR)
+	if err != nil {
+		return err
+	}
+	etcdJson, err := yaml.YAMLToJSON(etcdOperatorCRData)
+	if err != nil {
+		return err
+	}
+	err = json.Unmarshal(etcdJson, etcdOperatorCRStructure)
```
Comment (alaypatel07) [inline pkg/cmd/render/render.go:398]: You can probably use the `unstructured`[1] library to make this simpler/closer to the pattern seen around.

```
etcd := &unstructured.Unstructured{}
err := etcd.UnmarshalJSON([]byte(etcdJson))
...
err := unstructured.SetNestedField(etcd, true, "spec", "unsupportedConfigOverrides", "UseUnsupportedUnsafeNonHANonProductionUnstableEtcd")
...
// get json with etcd.MarshalJSON 
```
1. https://github.com/kubernetes/apimachinery/blob/15d95c0b2af3f4fcf46dce24105e5fbb9379af5a/pkg/apis/meta/v1/unstructured/helpers.go#L208:6

File: pkg/cmd/render/render.go [hunk 2]
```diff
@@ -377,6 +382,42 @@ func (t *TemplateData) setPlatform(infraConfigFilePath string) error {
 	return nil
 }
 
+func (t *TemplateData) setUnsupportedKeyForCR(etcdOperatorCR string) error {
+	type UnSupported struct {
+		UseUnsupportedUnsafeNonHANonProductionUnstableEtcd bool `json:"useUnsupportedUnsafeNonHANonProductionUnstableEtcd"`
+	}
+	etcdOperatorCRStructure := &operatorv1.MyOperatorResource{}
```
Comment (alaypatel07) [inline pkg/cmd/render/render.go:389]: in general, we use the variable name `etcd`[1] to refer to etcdOperatorCR, if possible drop the `OperatorCR` from variable names.

1. https://github.com/openshift/cluster-etcd-operator/blob/d2aa64e2e2e0388255cac2a8c648b37cbb7c1e26/pkg/operator/bootstrapteardown/waitforceo.go#L41

File: pkg/cmd/render/render.go [hunk 3]
```diff
@@ -377,6 +382,42 @@ func (t *TemplateData) setPlatform(infraConfigFilePath string) error {
...
+		return err
+	}
+	etcdOperatorCRStructure.Spec.UnsupportedConfigOverrides = runtime.RawExtension{Raw: b}
+
+	updateEtcdData, err := json.MarshalIndent(etcdOperatorCRStructure, "", "  ")
+	if err != nil {
+		return err
+	}
+
+	dstFile := fmt.Sprintf("/usr/share/bootkube/manifests/manifests/%s", filepath.Base(etcdOperatorCR))
```
Comment (alaypatel07) [inline pkg/cmd/render/render.go:414]: not sure how this would work, I think the idea is to write a new manifest file in an output directory[1] and then override it in bootkube by a cp command to the global manifest directory.

tldr: if you can output it in assets-output[2] directory, the installer will copy the manifest in the right place[1]

1. https://github.com/openshift/installer/blob/75738a342c1973121eedda7d91096d21c19194c9/data/data/bootstrap/files/usr/local/bin/bootkube.sh.template#L139
2. https://github.com/openshift/installer/blob/75738a342c1973121eedda7d91096d21c19194c9/data/data/bootstrap/files/usr/local/bin/bootkube.sh.template#L132

File: pkg/cmd/render/render.go [hunk 4]
```diff
@@ -221,6 +221,10 @@ func newTemplateData(opts *renderOpts) (*TemplateData, error) {
 		return nil, err
 	}
 
+	if templateData.Platform == "Libvirt" {
+		templateData.setUnsupportedKeyForCR("/manifests/0000_12_etcd-operator_01_operator.cr.yaml")
```
Comment (cfergeau) [inline pkg/cmd/render/render.go:225]: I don't understand why all libvirt clusters are assumed to need the etcd workaround for clusters with less than three nodes? There is nothing in the libvirt support which mandates this, is there?
